core/braintree/src/braintree_customer.erl

```diff
@@ -228,17 +228,31 @@ update_card(Customer, Card) ->
     %% Make card as default /after/ updating subscriptions: this way
     %%  subscriptions are not attached to a deleted card and thus do not
     %%  get cancelled before we can update their payment token.
-    NewCards = [braintree_card:update(
-                  braintree_card:make_default(NewCard, 'true')
-                 )
-               ],
-
-    %% Delete previous cards /after/ changing subscriptions' payment token.
-    lists:foreach(fun braintree_card:delete/1, OldCards),
-
-    UpdatedCustomer#bt_customer{credit_cards = NewCards
-                                ,subscriptions = NewSubscriptions
-                               }.
+    NewCard1 = braintree_card:update(
+                   braintree_card:make_default(NewCard, 'true')
+               ),
+
+    %% Delete previous cards and addresses /after/ changing subscriptions' payment token.
+    delete_old_cards_and_addresses(OldCards, NewCard1),
+
+    %%get all the new user info
+    braintree_customer:find(UpdatedCustomer).
+
+-spec delete_old_cards_and_addresses(bt_cards(), bt_card()) -> ok.
+delete_old_cards_and_addresses(OldCards, #bt_card{billing_address_id=NewAddressId}) ->
+    lists:foreach(fun(#bt_card{billing_address=OldAddress, billing_address_id=OldAddressId}=OldCard) ->
```

Comment:
I would prefer matching `OldAddressId` in the function head instead of a case here. I believe you'll need to do a guard for the `NewAddressId` clause.

```erlang
lists:foreach(
    fun(#bt_card{billing_address_id='undefined'}=OldCard) ->
            braintree_card:delete(OldCard);
        (#bt_card{billing_address_id=BillingAddressId}=OldCard) when BillingAddressId =:= NewAddressId ->
            braintree_card:delete(OldCard);
        (#bt_card{billing_address=OldAddress}=OldCard) ->
            braintree_card:delete(OldCard),
            braintree_address:delete(OldAddress)
        end
       ,OldCards
)
```

Something like that. Removes a nested chunk and I think captures the intent better.